diff --git a/tests/modetest/BUILD.gn b/tests/modetest/BUILD.gn
new file mode 100644
index 0000000..cb5a674
--- /dev/null
+++ b/tests/modetest/BUILD.gn
@@ -0,0 +1,37 @@
+import("//build/ohos.gni")
+
+ohos_executable("modetest") {
+
+  sources = [
+    "buffers.c",
+    "cursor.c",
+    "modetest.c",
+  ]
+
+  cflags = [
+      "-Wno-pointer-arith",
+  ]
+
+  include_dirs = [
+    "../",
+    ".",
+  ]
+  
+  configs = [ "//third_party/libdrm:libdrm_config" ]
+
+  public_configs = [ "//third_party/libdrm:libdrm_public_config" ]
+
+  deps = [
+    "//third_party/libdrm:libdrm",
+    "//third_party/libdrm/tests/util/:util",
+  ]
+
+  public_deps = []
+
+  install_images = [
+    "system",
+    "updater",
+  ]
+  part_name = "graphic_standard"
+  subsystem_name = "graphic"
+}
diff --git a/tests/util/BUILD.gn b/tests/util/BUILD.gn
new file mode 100644
index 0000000..8d7eddd
--- /dev/null
+++ b/tests/util/BUILD.gn
@@ -0,0 +1,27 @@
+import("//build/ohos.gni")
+
+ohos_static_library("util") {
+
+  sources = [
+    "format.c",
+    "kms.c",
+    "pattern.c",
+  ]
+
+  cflags = []
+
+  include_dirs = [
+    "../",
+    ".",
+  ]
+  
+  configs = [ "//third_party/libdrm:libdrm_config" ]
+
+  public_configs = [ "//third_party/libdrm:libdrm_public_config" ]
+
+  deps = [
+    "//third_party/libdrm:libdrm",
+  ]
+
+  public_deps = []
+}
diff --git a/tests/util/pattern.c b/tests/util/pattern.c
index bf1797d..f779517 100644
--- a/tests/util/pattern.c
+++ b/tests/util/pattern.c
@@ -985,7 +985,7 @@ static void fill_tiles_rgb16fp(const struct util_format_info *info, void *mem,
 			       unsigned int stride)
 {
 	const struct util_rgb_info *rgb = &info->rgb;
-	void *mem_base = mem;
+	//void *mem_base = mem;
 	unsigned int x, y;
 
 	/* TODO: Give this actual fp16 precision */
